File: 2025/12/03/structured-outputs-llms-reliable-json/typescript/src/repair.ts
/**
 * Auto-repair common JSON issues.
 */

export function repairJson(text: string): Record<string, any> | any[] | null {
  if (!text || !text.trim()) {
    return null;
  }
  
  // Remove comments (simple cases)
  // Remove // comments
  text = text.replace(/\/\/.*$/gm, '');
  // Remove /* */ comments
  text = text.replace(/\/\*.*?\*\//gs, '');
  
  // Remove trailing commas before } or ]
  text = text.replace(/,(\s*[}\]])/g, '$1');
  
  // Try to fix single quotes to double quotes (very simple cases only)
  // Match: 'key': 'value' or 'key': "value"
  text = text.replace(/'([^']*)':\s*/g, '"$1": ');
  // Match: "key": 'value'
  text = text.replace(/:\s*'([^']*)'/g, ': "$1"');
  
  // Trim whitespace
  text = text.trim();
  
  // Try parsing
  try {
    return JSON.parse(text);
  } catch {
    return null;
  }
}

export function fixTrailingComma(text: string): string {
  return text.replace(/,(\s*[}\]])/g, '$1');
}

export function removeComments(text: string): string {
  // Remove // comments
  text = text.replace(/\/\/.*$/gm, '');
  // Remove /* */ comments
  text = text.replace(/\/\*.*?\*\//gs, '');
  return text;
}

File: 2025/12/03/structured-outputs-llms-reliable-json/python/src/repair.py
"""Auto-repair common JSON issues."""
import json
import re
from typing import Any


def repair_json(text: str) -> dict | list | None:
    """
    Attempt to repair common JSON issues without retrying the LLM.
    
    Handles:
    - Single quotes to double quotes (simple cases)
    - Trailing commas
    - Comments (simple cases)
    - Extra whitespace
    
    Args:
        text: Potentially broken JSON string
    
    Returns:
        Repaired JSON object/array or None if repair fails
    """
    if not text or not text.strip():
        return None
    
    # Remove comments (simple cases)
    # Remove // comments
    text = re.sub(r'//.*?$', '', text, flags=re.MULTILINE)
    # Remove /* */ comments
    text = re.sub(r'/\*.*?\*/', '', text, flags=re.DOTALL)
    
    # Remove trailing commas before } or ]
    text = re.sub(r',(\s*[}\]])', r'\1', text)
    
    # Try to fix single quotes to double quotes (very simple cases only)
    # This is risky, so we only do it for obvious cases
    # Match: 'key': 'value' or 'key': "value"
    text = re.sub(r"'([^']*)':\s*", r'"\1": ', text)
    # Match: "key": 'value'
    text = re.sub(r":\s*'([^']*)'", r': "\1"', text)
    
    # Trim whitespace
    text = text.strip()
    
    # Try parsing
    try:
        return json.loads(text)
    except json.JSONDecodeError:
        return None


def fix_trailing_comma(text: str) -> str:
    """Remove trailing commas from JSON."""
    return re.sub(r',(\s*[}\]])', r'\1', text)


def remove_comments(text: str) -> str:
    """Remove comments from JSON-like text."""
    # Remove // comments
    text = re.sub(r'//.*?$', '', text, flags=re.MULTILINE)
    # Remove /* */ comments
    text = re.sub(r'/\*.*?\*/', '', text, flags=re.DOTALL)
    return text

